File: Barchart.BinarySerializer/Types/BinarySerializerDouble.cs
#region Using Statements

using Barchart.BinarySerializer.Buffers;

#endregion

namespace Barchart.BinarySerializer.Types;

/// <summary>
///     Reads (and writes) double values to (and from) a binary data source.
/// </summary>
public class BinarySerializerDouble : IBinaryTypeSerializer<double>
{
    #region Constants

    private const int ENCODED_LENGTH_IN_BYTES = sizeof(double);

    #endregion

    #region Methods

    /// <inheritdoc />
    public void Encode(IDataBufferWriter writer, double value)
    {
        writer.WriteBytes(BitConverter.GetBytes(value));
    }

    /// <inheritdoc />
    public double Decode(IDataBufferReader reader)
    {
        return BitConverter.ToDouble(reader.ReadBytes(ENCODED_LENGTH_IN_BYTES));
    }

    /// <inheritdoc />
    public bool GetEquals(double a, double b)
    {
        return a.Equals(b);
    }

    #endregion
}

File: typescript/src/types/binary-serializer-double.ts
import { DataReader } from "../buffers/data-reader.interface";
import { DataWriter } from "../buffers/data-writer.interface";
import { BinaryTypeSerializer } from "./binary-type-serializer.interface";

/**
 * Reads (and writes) double values to (and from) a binary data source.
 *
 * @public
 * @exported
 * @implements {BinaryTypeSerializer<number>}
 */
export class BinarySerializerDouble implements BinaryTypeSerializer<number> {
    get sizeInBytes(): number {
        return 8;
    }

    encode(writer: DataWriter, value: number): void {
        const buffer = new ArrayBuffer(this.sizeInBytes);
        new DataView(buffer).setFloat64(0, value, true);
        
        writer.writeBytes(new Uint8Array(buffer));
    }

    decode(reader: DataReader): number {
        const valueBytes = reader.readBytes(this.sizeInBytes);

        return new DataView(valueBytes.buffer, valueBytes.byteOffset, valueBytes.byteLength).getFloat64(0, true);
    }

    getEquals(a: number, b: number): boolean {
        return a === b;
    }
}